### ISSUE TITLE
Tracking Issue for `type_alias_type` feature flag

### ISSUE THREAD

**Issue by GuillaumeGomez** [rust-lang/rust#112792]
To be able to fully use https://github.com/rust-lang/rust/pull/108860 and not limit it to opaque types only.

**Comment by orhid** [comment]
i think i found a bug with this feature, sorry if this is not the right place to report this. the following code compiles without the feature enabled, while with it the second to last line throws the `error[E0207]: the type parameter 'A' is not constrained by the impl trait, self type, or predicates` error:

```rust
#![feature(lazy_type_alias)]

struct Foo<A,B> {
    foo_a: A,
    foo_b: B,
}

type Faa<A> = Foo<A,A>;

// this one does not
impl<A> Faa<A> {}

// this one compiles
impl<A> Foo<A,A> {}
```

[link to playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=4daea394dc0b9fe4559c218745367460)

**Comment by fmease** [comment]
@orhid Thanks for your report! Usually you just open a separate issue since a tracking issue is meant to record the overall progress on a feature and to be a hub for connecting to issues. No worries though, the issue description is a bit terse right now, I'm gonna update it later. I've opened #116100 for your issue.

**Comment by CAD97** [comment]
Recording:

Trailing `where` clauses on top level type aliases are currently syntax-stable (i.e. allowed under an inactive `#[cfg]`), and is only feature gated semantically (i.e. after macro expansion).

For other unstable features that were accidentally allowed as syntax-stable (e.g. `macro` items), we now have the forward-compatibility warning `unstable_syntax_pre_expansion` (https://github.com/rust-lang/rust/issues/65860). We *could* potentially add this syntax into that machinery if we want to be conservative w.r.t. the unstable feature.

But on the other hand, trailing `where` is already allowed for associated types, so it being syntax stable for top level type aliases isn't particularly worrying or risky.

TL;DR: it being syntax stable deserves to be recorded, and is fine to remain such without any FCW (imo).

**Comment by fmease** [comment]
> But on the other hand, trailing where is already allowed for associated types, so it being syntax stable for top level type aliases isn't particularly worrying or risky.

@CAD97, I might've misunderstood but to me it seems like you think that trailing where-clauses are deprecated while it's actually *leading* where-clauses that are deprecated.

Trailing where-clauses are the new default for both associated types and lazy type aliases. Leading WCs on assoc tys are warned against and rejected on lazy ty aliases during semantic analysis. See #114662.

I forgot what the specific reason was for continuing to allow leading WCs syntactically but backward compatibility is likely the answer. See #89122 for more information.

**Comment by CAD97** [comment]
For clarity, I was aware this was the case, and was just noting the syntactic stability had preceded the semantic stability in this case (and concluded this is fine given the stability of the same syntax for associated types). The mention of FCW is due to expectation but merely that it *is* unstable and *could* change.
